### PR TITLE
Added custom events click, hover and focus

### PR DESCRIPTION
**Pull request by chaosmail**
Hi n3-chart team,

I added 3 custom events to the chart directive. Due to the glass g-element displayed for the scrubber tooltip, I was not able to implement *click* and *hover* in scrubber mode.

* *click*: the mouse clicks on a dot or column (tooltip modes: axes, none)
* *hover*: the mouse hovers over a dot or column (tooltip modes: axes, none)
* *focus*: an event that triggers the scrubber focus (tooltip mode: scrubber)

Every event handler is called with the data element *d* and the position *i* in the array of values of a series. 

I used this syntax for now, let me know if we can find something better:

```html
<linechart data="data" options="options" click="onClick" hover="onHover" focus="onFocus"></linechart>
```

```js
$scope.onClick = function(d, i){ ... };
$scope.onHover = function(d, i){ ... };
$scope.onFocus = function(d, i){ ... };
```

Here is an example for *click* and *hover*: [http://plnkr.co/edit/PTxVo3qKEnooK1PcsXMX](http://plnkr.co/edit/PTxVo3qKEnooK1PcsXMX?p=preview)

Here is an example for *focus*: [http://plnkr.co/edit/vUOa1v4UFRlbPZINCZ6X](http://plnkr.co/edit/vUOa1v4UFRlbPZINCZ6X?p=preview)

Best,
Christoph

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -135,6 +135,48 @@ This is more a hack. The chart usually tries to infer its own dimensions regardi
...
 
+#### Custom events
+We can attach event handlers for *click*, *hover* and *focus* events of the chart.
+
+##### Click
+The event handler for the *click* event get's triggered when the mouse clicks on a dot or column of the chart in tooltip modes *none* and *axes*.
+
+```js
+$scope.onClick = function(d, i){
+  conosole.log(d, i);
```
Comment (lorem--ipsum) [inline README.md]: Unfortunately, `conosole` is not well-supported. Let's use `console` instead, if you don't mind :-)

File: README.md [hunk 2]
```diff
@@ -135,6 +135,48 @@ This is more a hack. The chart usually tries to infer its own dimensions regardi
...
+```html
+<linechart data="data" options="options" click="onClick"></linechart>
+```
+
+##### Hover
+The event handler for the *hover* event get's triggered when the mouse hovers over a dot or column of the chart in tooltip modes *none* and *axes*.
+
+```js
+$scope.onHover = function(d, i){
+  conosole.log(d, i);
```
Comment (lorem--ipsum) [inline README.md]: Same.

File: README.md [hunk 3]
```diff
@@ -135,6 +135,48 @@ This is more a hack. The chart usually tries to infer its own dimensions regardi
...
+```html
+<linechart data="data" options="options" hover="onHover"></linechart>
+```
+
+##### Focus
+The event handler for a *focus* event get's triggered when the mouse hovers over the chart in tooltip mode *scrubber*.
+
+```js
+$scope.onFocus = function(d, i){
+  conosole.log(d, i);
```
Comment (lorem--ipsum) [inline README.md]: Same.

File: src/line-chart.coffee [hunk 4]
```diff
@@ -9,6 +9,7 @@ directive('linechart', ['n3utils', '$window', '$timeout', (n3utils, $window, $ti
   link  = (scope, element, attrs, ctrl) ->
     _u = n3utils
     dim = _u.getDefaultMargins()
+    dispatch = _u.getEventDispatcher()
```
Comment (lorem--ipsum) [inline src/line-chart.coffee]: Just out of curiosity, is there any specific reason you didn't use the `initialHandlers` object ?

File: src/utils/scrubber.coffee [hunk 5]
```diff
@@ -48,6 +48,8 @@
 
           v = that.getClosestPoint(series.values, axes.xScale.invert(x))
 
+          dispatch.focus(v, series.values.indexOf(v))
```
Comment (lorem--ipsum) [inline src/utils/scrubber.coffee]: We could probably support clicks on the glass in scrubber mode, with some sort of indication passed along the data about the distance from data points...

File: test/unit/events.mocha.coffee [hunk 6]
```diff
@@ -0,0 +1,208 @@
+describe 'event handling', ->
```
Comment (lorem--ipsum) [inline test/unit/events.mocha.coffee]: :+1:

**Comment by lorem--ipsum** [comment]
Very good PR, thank you for your contribution.

**Comment by chaosmail** [comment]
d3.events are very nice, clean and robust for dispatching (a lot of) events, maybe the initialHandler can be replaced with d3 events at one point.

**Comment by chaosmail** [comment]
The glass layer somehow blocks the click events, so that the dots beneath don't get triggered. There is for sure also an easy way to propagate the click event through this layer - but I haven't looked into this yet.

**Comment by chaosmail** [comment]
Yeah, and the point about the scrubber mode sounds good, let me add a third parameter to the callback of the focus event

**Comment by chaosmail** [comment]
what do you think about for sending an array containing the inverted x and y positions from the mouse cursor?

```js
xInvert = axes.xScale.invert(x)
yInvert = axes.yScale.invert(y)

dispatch.focus(v, series.values.indexOf(v), [xInvert, yInvert])
```

**Comment by chaosmail** [comment]
I updated the demo, actually this additional value can be very useful too. [http://plnkr.co/edit/vUOa1v4UFRlbPZINCZ6X?p=preview](http://plnkr.co/edit/vUOa1v4UFRlbPZINCZ6X?p=preview)

**Comment by chaosmail** [comment]
Fixes #199
Fixes #211

**Comment by lorem--ipsum** [comment]
:poultry_leg:

**Comment by chaosmail** [comment]
Seems like the linked issues #199 and #211 are not automatically closed

**Comment by dorlandode** [comment]
Is there a reason the click event an hover events only work on tooltip mode axes and none?  I was hoping it work on scrubber as well.

**Comment by chaosmail** [comment]
Hi @dorlandode 
the reason for this is how the scrubber tooltip works (it generates a transparent rectangle in front of all elements) which makes it hard to enable this events in scrubber mode.
However, we are actively developing version 2 (a complete rewrite in TypeScript) where this problem will be solved -  and all events will be available for all tooltips.
Best,
Christoph

**Comment by dorlandode** [comment]
awesome, thanks for the update.  What is the ETA on version 2?

**Pull request merged**